File: hard/string_permutations/string.permutations.py
# Reuben Thorpe (2016), CodeEval [String Permutations v1.0]

from sys import argv
from itertools import permutations as perm
from math import factorial


def sort(fileName):

    for line in open(fileName, 'r'):
        n = len(line.strip())
        limit = factorial(n)
        digits = sorted((char for char in line if char.isdigit()))
        alphaUpper = sorted((char for char in line if char.isupper()))
        alphaLower = sorted((char for char in line if char.islower()))

        for i, seq in enumerate(perm(digits + alphaUpper + alphaLower)):
            print("".join(seq), end="")
            if i < limit-1:
                print(",", end="")
            else:
                print()


if __name__ == "__main__":
    sort(argv[1])

File: hard/string_permutations/string.permutations.cpp
// Reuben Thorpe (2016), CodeEval [String Permutations v1.1]
#include <iostream>
#include <algorithm>
#include <fstream>

using namespace std;

int fac(int n) {
  // Compute factorial
  return (n == 1 || n == 0) ? 1 : fac(n - 1) * n;
}


int main(int argc, char* argv[]) {
  ifstream inFile(argv[1]);
  string line;

  while (inFile >> line) {
    int i = 0;
    int limit = fac(line.length());
    sort(line.begin(), line.end());

    do {
      cout << line << ',';
      i++;
    } while (next_permutation(line.begin(), line.end()) && i < limit-1);

    next_permutation(line.end(), line.begin());
    cout << line << '\n';
  }

  return(0);
}
